### PR TITLE
Add unit type in number detection

### PR DESCRIPTION
**Pull request by amansrivastava17**
**Done following changes:**
1. Create a new key `PARAMETER_NUMBER_UNIT_TYPE` in _get_parameters_dictionary_ method to capture unit type passed in request.

2. Passed unit type in NumberDetector constructor
3. Added new column `unit_type` in _units.csv_.
4. Changes language_number_detector.units_map to NamedTuple to store unit value and unit type.
5. Added validation condition in main number detector wrapper to check for unit type, below is the condition
```    validated_number, validated_number_text = [], []
        for number_value_dict, original_text in zip(number_data[0], number_data[1]):
            number_value = number_value_dict[NUMBER_DETECTION_RETURN_DICT_VALUE]
            number_unit = number_value_dict[NUMBER_DETECTION_RETURN_DICT_UNIT]
            if self.min_digit <= len(number_value) <= self.max_digit:
                if self.unit_type and (number_unit is None or
                                       self.language_number_detector.units_map[number_unit].type != self.unit_type):
                    continue
                validated_number.append(number_value_dict)
                validated_number_text.append(original_text)
```

### CODE REVIEW

File: ner_v2/detectors/numeral/number/number_detection.py [hunk 1]
```diff
@@ -127,7 +130,13 @@ def detect_entity(self, text, **kwargs):
         number_data = self.language_number_detector.detect_number(self.processed_text)
         validated_number, validated_number_text = [], []
         for number_value_dict, original_text in zip(number_data[0], number_data[1]):
-            if self.min_digit <= len(number_value_dict[NUMBER_DETECTION_RETURN_DICT_VALUE]) <= self.max_digit:
+            number_value = number_value_dict[NUMBER_DETECTION_RETURN_DICT_VALUE]
+            number_unit = number_value_dict[NUMBER_DETECTION_RETURN_DICT_UNIT]
+            if self.min_digit <= len(number_value) <= self.max_digit:
+                if self.unit_type and (not self.language_number_detector.units_map.get(number_unit) or
+                                       (self.language_number_detector.units_map.get(number_unit) and
+                                        self.language_number_detector.units_map[number_unit].type != self.unit_type)):
```
Comment (chiragjn) [inline ner_v2/detectors/numeral/number/number_detection.py]: Why not pass on the unit type to the lower level detectors and exclude irrelevant units from the regex itself? Would make the detection faster.

Comment (chiragjn) [reply ner_v2/detectors/numeral/number/number_detection.py]: Why is this needed? not self.language_number_detector.units_map.get(number_unit)

File: ner_v2/detectors/numeral/number/en/data/units.csv [hunk 2]
```diff
@@ -1,3 +1,3 @@
-unit_value,unit_variants
-rupees,rupees | rupee | rs | rupya | rupaya | rupaye | rupye | rupay | paisa | paise | inr | ₹  
-dollar,Dollar | usd | $
+unit_type,unit_value,unit_variants
```
Comment (chiragjn) [inline ner_v2/detectors/numeral/number/en/data/units.csv]: readme needs to be updated about this

File: ner_v2/detectors/numeral/number/number_detection.py (hunk 1 above)
Comment (amansrivastava17) [reply ner_v2/detectors/numeral/number/number_detection.py]: If case no unit is present in number detected but unit_type is passed in request.

File: ner_v2/detectors/numeral/number/README.md [hunk 3]
```diff
@@ -109,12 +109,12 @@ Below is the brief about how to create data files `numerals_constant.csv`, `unit
...
-   | :--------: | :----------------------------------------------------------: |
-   |   rupees   | rupees \| rupay \| paisa \| paise \| inr \| रूपीस \| रुपया \| रूपए\| पैसा\| पैसे\| ₹ |
-   |   dollar   |                  Dollar \| usd \| डॉलर \| $                  |
+   | unit_type | unit_value |                        unit_variants                         |
+   | --------- | :--------: | :----------------------------------------------------------: |
+   | currency  |   rupees   | rupees \| rupay \| paisa \| paise \| inr \| रूपीस \| रुपया \| रूपए\| पैसा\| पैसे\| ₹ |
+   | currency  |   dollar   |                  Dollar \| usd \| डॉलर \| $                  |
 
-   Here, 1st column will contain the value of unit which will be return by detector module, while 2nd column contain all the possible variants of units in language and english script.
+   Here, 1st column have the type of unit variants it have correspoding to unit variants, 2nd contain the value of unit which will be return by number detector and 3rd column contains all the possible variants of units in language and english script.
```
Comment (chiragjn) [inline ner_v2/detectors/numeral/number/README.md]: Phrasing:
Here, the 1st column contains the type of unit (E.g. dollars, euros are "currency",  centimetre, metre, kilometre are "distance"), 2nd column contains the value of unit which will be returned by number detector and 3rd column contains all the possible variants of that unit value (delimited by pipe  `|`) for the language you are adding. (It is recommended to add Romanised versions of the variants you are adding)

File: ner_v2/detectors/numeral/number/standard_number_detector.py [hunk 4]
```diff
@@ -16,7 +16,7 @@
 
 
 class BaseNumberDetector(object):
-    def __init__(self, entity_name, data_directory_path):
+    def __init__(self, entity_name, data_directory_path, unit_type=None):
```
Comment (chiragjn) [inline ner_v2/detectors/numeral/number/standard_number_detector.py]: Incomplete docs,
missing, entity_name and unit_type

File: ner_v2/detectors/numeral/number/README.md [hunk 5]
```diff
@@ -134,8 +134,11 @@ class NumberDetector(BaseNumberDetector):
     data_directory_path = os.path.join((os.path.dirname(os.path.abspath(__file__)).rstrip(os.sep)),
                                        LANGUAGE_DATA_DIRECTORY)
 
-    def __init__(self, entity_name):
-        super(NumberDetector, self).__init__(entity_name=entity_name,                                      data_directory_path=NumberDetector.data_directory_path)
+    def __init__(self, entity_name, unit_type=None):
+        super(NumberDetector, self).__init__(entity_name=entity_name,
+                                             unit_type=unit_type,
```
Comment (chiragjn) [inline ner_v2/detectors/numeral/number/README.md]: Move unit_type to the end, after `data_directory_path` as `data_directory_path` is a required argument

File: ner_v2/detectors/numeral/number/en/number_detection.py [hunk 6]
```diff
@@ -10,8 +10,9 @@ class NumberDetector(BaseNumberDetector):
     data_directory_path = os.path.join((os.path.dirname(os.path.abspath(__file__)).rstrip(os.sep)),
                                        LANGUAGE_DATA_DIRECTORY)
 
-    def __init__(self, entity_name):
+    def __init__(self, entity_name, unit_type=None):
```
Comment (chiragjn) [inline ner_v2/detectors/numeral/number/en/number_detection.py]: Add a default `entity_name='number'`

File: ner_v2/detectors/numeral/number/README.md [hunk 7]
```diff
@@ -134,8 +134,11 @@ class NumberDetector(BaseNumberDetector):
     data_directory_path = os.path.join((os.path.dirname(os.path.abspath(__file__)).rstrip(os.sep)),
                                        LANGUAGE_DATA_DIRECTORY)
 
-    def __init__(self, entity_name):
-        super(NumberDetector, self).__init__(entity_name=entity_name,                                      data_directory_path=NumberDetector.data_directory_path)
+    def __init__(self, entity_name, unit_type=None):
```
Comment (chiragjn) [inline ner_v2/detectors/numeral/number/README.md]: Add a default `entity_name='number'`

File: ner_v2/detectors/numeral/number/standard_number_detector.py [hunk 8]
```diff
@@ -299,5 +302,7 @@ def _update_processed_text(self, original_number_list):
 
 
 class NumberDetector(BaseNumberDetector):
-    def __init__(self, entity_name, data_directory_path):
-        super(NumberDetector, self).__init__(entity_name=entity_name, data_directory_path=data_directory_path)
+    def __init__(self, entity_name, data_directory_path, unit_type):
```
Comment (chiragjn) [inline ner_v2/detectors/numeral/number/standard_number_detector.py]: Add a default `entity_name='number'`

Comment (chiragjn) [inline ner_v2/detectors/numeral/number/standard_number_detector.py]: Add a default `unit_type=None`

File: ner_v2/detectors/numeral/number/README.md [hunk 9]
```diff
@@ -198,10 +201,12 @@ class NumberDetector(BaseNumberDetector):
     data_directory_path = os.path.join((os.path.dirname(os.path.abspath(__file__)).rstrip(os.sep)),
                                        LANGUAGE_DATA_DIRECTORY)
 
-    def __init__(self, entity_name):
+    def __init__(self, entity_name, unit_type=None):
         super(NumberDetector, self).__init__(entity_name=entity_name,
+                                             unit_type=unit_type,
```
Comment (chiragjn) [inline ner_v2/detectors/numeral/number/README.md]: Move `unit_type` to the end, after `data_directory_path` as `data_directory_path` is a required argument

**Pull request merged**